File: search_in_unknown_length.py
# """
# This is ArrayReader's API interface.
# You should not implement it, or speculate about its implementation
# """
#class ArrayReader:
#    def get(self, index: int) -> int:

#Time Complexity  : O(logn)
# Space Complexity : O(1)

class Solution:
    def search(self, reader, target):
        """
        :type reader: ArrayReader
        :type target: int
        :rtype: int
        """
        
        start = 0
        end  = 1
        
        # Iterate through the elements and get the range where the target is closer by moving the start and end based on target value
        while ( reader.get(end) < target) :
            start = end 
            end = end * 2
        
        # do binary search for the last elements that contains the target
        while (start <=  end) :
            mid = start + (end - start)// 2
            
            if reader.get(mid) ==  target :
                return mid
            
            elif reader.get(mid) > target :
                end = mid - 1
            else :
                start = mid + 1
                
        
        return -1
                
        

File: SearchInUnknownLength.java
// Time Complexity : O(log(m*n))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : YES
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach in three sentences only

/**
 * We keep on increasing the distance between low and high by twice, so the TC
 * for this operation will be O(log m).
 * After finding the range in which our element lies, for searching in that range will be O(log n)
 */
public class SearchInUnknownLength {
    interface ArrayReader {
        int get(int index);
    }

    public int search(ArrayReader reader, int target) {
        if(reader.get(0) == target) return 0;

        int low = 0;
        int high = 1;

        while(reader.get(high) < target) {
            low = high;
            high = 2*high;
        }

        while(low <= high) {
            int mid = (high-low)/2 + low;

            if(target == reader.get(mid)) return mid;
            else if (target < reader.get(mid)) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return -1;

    }
}
